### ISSUE TITLE
Conflicts with Next.js SSR (hydration errors with non-deterministic initial state)

### ISSUE THREAD

**Issue by skyhancloud** [NickvanDyke/eslint-plugin-react-you-might-not-need-an-effect#33]
Hi 👋

I ran into a case where following the recommendation from `react-you-might-not-need-an-effect/no-initialize-state` causes a hydration error in Next.js when using non-deterministic initial state.

### Example

```tsx
const [height, setHeight] = useState(() => generateRandomHeight());

useEffect(() => {
  const interval = setInterval(() => setHeight(generateRandomHeight()), 1500);
  return () => clearInterval(interval);
}, []);
```

This works fine in a client-only React app. But in a Next.js SSR app, hydration breaks because `generateRandomHeight()` returns different values between server render and client render.

The hydration error:

```
A tree hydrated but some attributes of the server rendered HTML didn't match the client properties.
```
### Next.js safe approach

To avoid hydration mismatches, I need to use a deterministic initial value (e.g. 0) and then update inside a `useEffect`:

```tsx
const [height, setHeight] = useState(0);

useEffect(() => {
  setHeight(generateRandomHeight());

  const interval = setInterval(() => setHeight(generateRandomHeight()), 1500);
  return () => clearInterval(interval);
}, []);
```

This eliminates the hydration error, but now the linter warns:

```
Avoid initializing state in an effect. Instead, pass "height"'s initial value to its `useState`
```

### Question

How should this rule behave in SSR contexts like Next.js?

* In client-only React, the rule makes sense.
* In SSR apps, initializing non-deterministic state in `useState` is not possible.

Should the rule:

1. Document this SSR caveat,
2. Provide a way to disable/ignore the rule in these cases,
3. Or detect common SSR frameworks and adjust behavior?

Thanks for the great work on this plugin! 🙏

**Comment by NickvanDyke** [comment]
Yeah that might be worth it, although it's went this long without an SSR issue haha, and this issue now exists as a reference. 🤔 

Unfortunately I don't think the rule can detect when it's in a NextJS project, because ESLint runs on one file at a time. I see the [use client](https://nextjs.org/docs/app/api-reference/directives/use-client) directive exists which I could check for, but if I understand correctly, it only needs to exist at the server-client component boundary, and not necessarily on every client-side component.

I would prefer to detect it in a way that "just works" for users though - you definitely know Nextjs better than me, so if you have any ideas to detect this on a per-file basis, I'm all ears!

Otherwise SSR users may just have to disable this rule in their ESLint config, or at each instance, depending on their tolerance. If it becomes a widespread annoyance then I could remove it from the recommended config.

**Comment by NickvanDyke** [comment]
> I don't have any experience developing plugins, so I don't know how good an idea or solution I could come up with

That's okay, I was more curious about your experience with Next.js and any giveaways within a file that the effect is used specifically client-side 🙂 But if there aren't any, then no big deal!

I will keep this in mind though, in case there's a way I can reduce friction for people/projects it affects, without sacrificing general utility 😁

**Comment by NickvanDyke** [comment]
Thanks @kurtextrem ! The rule's message now mentions that and links to that article. Closing this issue accordingly, but let me know if I missed anything.
